### PR TITLE
Adds Static Addresses

### PR DESCRIPTION
**Pull request by kluman**
Fixes Issue #25

### CODE REVIEW

File: src/main/java/gyro/google/compute/AbstractAddressResource.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    public String getAddress() {
+        return address;
+    }
+
+    public void setAddress(String address) {
+        this.address = address;
+    }
+
+    /**
+     * The prefix length if the resource reprensents an IP range.
```
Comment (jpencola) [inline src/main/java/gyro/google/compute/AbstractAddressResource.java]: typo: 'reprensents'

Comment (kluman) [reply src/main/java/gyro/google/compute/AbstractAddressResource.java]: Fixed.

File: src/main/java/gyro/google/compute/AddressFinder.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            }
+
+        } catch (GoogleJsonResponseException e) {
+            if (e.getDetails().getCode() == 404) {
+                return new ArrayList<>();
+            } else {
+                throw new GyroException(e.getDetails().getMessage());
+            }
+        }
+    }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressFinder.java]: This is handled by the parent class.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    public void copyFrom(Address model) {
+        setName(model.getName());
+        setDescription(model.getDescription());
+        setAddress(model.getAddress());
+        setPrefixLength(model.getPrefixLength());
+        setAddressType(model.getAddressType());
+        setPurpose(model.getPurpose());
+
+        String subNetworkLink = model.getSubnetwork();
+        if (subNetworkLink != null) {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractAddressResource.java]:
```suggestion
        setSubnetwork(null);
        if (subNetworkLink != null) {
```

File: src/main/java/gyro/google/compute/AbstractAddressResource.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            String sunbNetworkName = model.getSubnetwork().substring(model.getSubnetwork().lastIndexOf('/') + 1);
+
+            if (!"default".equals(sunbNetworkName)) {
+                setSubnetwork(findById(SubnetworkResource.class, sunbNetworkName));
+            }
+        }
+
+        String networkLink = model.getNetwork();
+        if (networkLink != null) {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractAddressResource.java]:
```suggestion
        setNetwork(null);
        if (networkLink != null) {
```

File: src/main/java/gyro/google/compute/AbstractAddressResource.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        setAddressType(model.getAddressType());
+        setPurpose(model.getPurpose());
+
+        String subNetworkLink = model.getSubnetwork();
+        if (subNetworkLink != null) {
+            String sunbNetworkName = model.getSubnetwork().substring(model.getSubnetwork().lastIndexOf('/') + 1);
+
+            if (!"default".equals(sunbNetworkName)) {
+                setSubnetwork(findById(SubnetworkResource.class, sunbNetworkName));
+            }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractAddressResource.java]: Should directly set `setSubnetwork(model.getSubnetwork())` as the getter of `self-link` of SubnetworkResource has the `@Id` annotation on it and not the name.

File: src/main/java/gyro/google/compute/AddressResource.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        if (error != null) {
+            throw new GyroException(error.toPrettyString());
+        }
+        refresh();
+    }
+
+    @Override
+    public void doDelete(GyroUI ui, State state) throws Exception {
+        Compute compute = createClient(Compute.class);
+        compute.addresses().delete(getProjectId(), getRegion(), getName()).execute();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressResource.java]: Should handle operation error.

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+        } catch (GoogleJsonResponseException e) {
+            if (e.getDetails().getCode() == 404) {
+                return new ArrayList<>();
+            } else {
+                throw new GyroException(e.getDetails().getMessage());
+            }
+        }
+    }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/GlobalAddressFinder.java]: Handled by the super class.

File: src/main/java/gyro/google/compute/GlobalAddressResource.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        if (error != null) {
+            throw new GyroException(error.toPrettyString());
+        }
+        refresh();
+    }
+
+    @Override
+    public void doDelete(GyroUI ui, State state) throws Exception {
+        Compute compute = createClient(Compute.class);
+        compute.globalAddresses().delete(getProjectId(), getName()).execute();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/GlobalAddressResource.java]: Should handle operation error.

File: src/main/java/gyro/google/compute/AddressFinder.java (hunk 2 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressFinder.java]: Fixed.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java (hunk 3 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AbstractAddressResource.java]: Done.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java (hunk 4 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AbstractAddressResource.java]: Done.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java (hunk 5 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AbstractAddressResource.java]: Done.

File: src/main/java/gyro/google/compute/AddressResource.java (hunk 6 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressResource.java]: Done.

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java (hunk 7 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/GlobalAddressFinder.java]: Done.

File: src/main/java/gyro/google/compute/GlobalAddressResource.java (hunk 8 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/GlobalAddressResource.java]: Done.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+     * Internal Google id.
+     */
+    @Output
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractAddressResource.java]: Never used.

File: src/main/java/gyro/google/compute/AddressFinder.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        String pageToken = null;
+
+        do {
+            AddressAggregatedList aggregatedList = client.addresses().aggregatedList(getProjectId())
+                .setPageToken(pageToken)
+                .execute();
+
+            if (aggregatedList.getItems() != null) {
+                aggregatedList.getItems().remove("global");
+                pageToken = aggregatedList.getNextPageToken();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressFinder.java]: This line can be taken out of the `if` block and the corresponding `else` block can go away along with the `break`.

File: src/main/java/gyro/google/compute/AddressFinder.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+                List<Address> aggregated = aggregatedList.getItems().values().stream()
+                    .map(AddressesScopedList::getAddresses)
+                    .filter(Objects::nonNull)
+                    .flatMap(List::stream)
+                    .collect(Collectors.toList());
+
+                if (!aggregated.isEmpty()) {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressFinder.java]: Is this check necessary ?

File: src/main/java/gyro/google/compute/AddressFinder.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+        } while (pageToken != null);
+
+        return addresses;
+    }
+
+    @Override
+    protected List<Address> findGoogle(Compute client, Map<String, String> filters) throws Exception {
+        if (filters.containsKey("region")) {
+            List<Address> addresses = new ArrayList<>();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressFinder.java]: This line can be taken out of the `if` block, then we will have a single return statement.

File: src/main/java/gyro/google/compute/AddressResource.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        Compute compute = createClient(Compute.class);
+        Address address = copyTo()
+            .setRegion(getRegion())
+            .setNetworkTier(getNetwork() != null ? getNetwork().getSelfLink() : null);
+        Operation.Error error = waitForCompletion(compute, compute.addresses().insert(getProjectId(), getRegion(), address).execute());
+
+        if (error != null) {
+            throw new GyroException(error.toPrettyString());
+        }
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AddressResource.java]:
```suggestion
        
        refresh();
```

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+        } while (pageToken != null);
+
+        return addresses;
+    }
+
+    @Override
+    protected List<Address> findGoogle(Compute client, Map<String, String> filters) throws Exception {
+        if (filters.containsKey("filter")) {
+            List<Address> addresses = new ArrayList<>();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/GlobalAddressFinder.java]: This line can be taken out of the `if` block then we can have a single return statement.

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            AddressList addressList = client.globalAddresses().list(getProjectId())
+                .setPageToken(pageToken)
+                .execute();
+            pageToken = addressList.getNextPageToken();
+
+            if (addressList.getItems() != null) {
+                addresses.addAll(addressList.getItems());
+            } else {
+                break;
+            }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/GlobalAddressFinder.java]: This `else` block is not necessary.

File: src/main/java/gyro/google/compute/AbstractAddressResource.java (hunk 9 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AbstractAddressResource.java]: Removed.

File: src/main/java/gyro/google/compute/AddressFinder.java (hunk 10 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressFinder.java]: Done.

File: src/main/java/gyro/google/compute/AddressFinder.java (hunk 12 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressFinder.java]: Done.

File: src/main/java/gyro/google/compute/AddressFinder.java (hunk 11 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressFinder.java]: Nope, removed.

File: src/main/java/gyro/google/compute/AddressResource.java (hunk 13 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/AddressResource.java]: Done.

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java (hunk 14 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/GlobalAddressFinder.java]: Done.

File: src/main/java/gyro/google/compute/GlobalAddressFinder.java (hunk 15 above)
Comment (kluman) [reply src/main/java/gyro/google/compute/GlobalAddressFinder.java]: Done.

**Pull request merged**